casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -411,4 +418,48 @@ object ProtoUtil {
     (missingParents union missingJustifications).toList
   }
 
+  // Return hashes of all blocks that are yet to be seen by the passed in block
+  def unseenBlockHashes[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMessage
+  ): F[Set[BlockHash]] =
+    for {
+      latestMessages        <- dag.latestMessages
+      latestMessagesOfBlock <- ProtoUtil.toLatestMessage(block.justifications, dag)
+      result <- latestMessages.toList
+                 .traverse {
+                   case (validator, latestMessage) =>
+                     val lastSeenBlockHash = latestMessagesOfBlock(validator)
+                     DagOperations
+                       .bfTraverseF(List(latestMessage))(
+                         block => getCreatorJustificationUnlessGoal(dag, block, lastSeenBlockHash)
+                       )
+                       .map(_.blockHash)
+                       .toSet
+                 }
+                 .map(_.flatten.toSet)
+    } yield result -- latestMessagesOfBlock.values.map(_.blockHash) - block.blockHash
+
+  private def getCreatorJustificationUnlessGoal[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMetadata,
+      goal: BlockMetadata
+  ): F[List[BlockMetadata]] =
```

Comment:
Why you returning `List` instead of `Option` here considering this function can only return 0 or 1 blocks? Is it because `DagOperations.bfTraverseF` expects `F[List[BlockMetadata]]`?

Reply:
See https://discordapp.com/channels/391435113415573504/410182072054054912/570082282359226388

Reply:
Sorry I missed the conversation. I think your best bet was implementing such function with `tailRecM` like this:
```scala
def kent[F[_]: Monad](
    start: BlockMetadata
  )(nextFunction: BlockMetadata => F[BlockMetadata])(until: BlockMetadata => Boolean): F[Set[BlockMetadata]] =
    Monad[F].tailRecM((start, Set(start))) {
      case (block, result) =>
        if (!until(block)) {
          for {
            x <- nextFunction(block)
          } yield (x, result + x).asLeft[Set[BlockMetadata]]
        } else {
          result.asRight[(BlockMetadata, Set[BlockMetadata])].pure[F]
        }
    }
```

I am fine with keeping the code as it is though if you want; just wanted to share my thoughts on this